### DISCUSSION TITLE
How to run semantic-release as a Docker command-line application?

### DISCUSSION THREAD

**Discussion by finn-christiansen** [semantic-release/semantic-release#1847] [Q&A] [answered]
I want to run semantic-release without the need to setup node locally. I want to run semantic release in docker but without executing `git pull` in the container.

desired functionality:

I want to run the following command inside a git-repository on any machine that has docker as single dependency:

```
docker run --rm -it \
    --env GH_TOKEN=xxx \
    --env REPOSITORY_URL=yyy \
    -v $(pwd)/.git:/usr/src/app/.git \
    my-semantic-release:latest
```

Dockerfile

```
FROM node:14-alpine

RUN apk add --no-cache git

WORKDIR /usr/src/app

RUN npm install semantic-release @semantic-release/exec

CMD npx semantic-release \
    --no-ci \
    --repository-url $REPOSITORY_URL
```

This works, but unfortunately, git modifies my `.git` folder in the Docker container. This leads to permission problems if I run docker locally after executing sematic-release inside the docker container.

I know that I can fix the permissions afterward with `sudo chown ... && sudo chmod ...` but this doesn't seem like a best practice solution. Additionally, I could simply make a copy of the `.git`-folder inside the container. This way, only the copy will be modified and my "local" `.git`-folder stays untouched. But this also seems not like a clean solution.

Is someone using a similar setup and knows a clean solution for this?

**Comment by bigbrozer** [comment]
You can add `USER_ID` and `GROUP_ID` as new ENV in the Dockerfile. Then, add an ENTRYPOINT which creates a user using these uid and gid and execute semantic-release as this user. When running the container at host, ensure you set `USER_ID` and `GROUP_ID` to the uid and gid of your *real* user (`docker run -e ...`). This will map the permissions on the host inside the container and avoid issues you are encountering.

**Comment by finn-christiansen** [comment, marked as answer]
Thank you @bigbrozer for the tips. I was able to solve my issue but there have been two problems:

- The default user id for the first non-root user for most Linux systems is "1000". Therefore I got a conflict because the "node" user of the Docker image had the user id "1000" and my local user, too.
- If I want to build the Docker image once in a while during a CI-pipeline and use it on different machines with different users, the adding of the user must take place during runtime.

I solved this problem with the following files and commands:

`Dockerfile`

```Dockerfile
FROM node:14-alpine

# Node has UID=1000 which is the default UID for ubuntu which can lead to conflicts.
# This image is intended to be run by the USER and UID of the host system, so the
# default node user is not required.
RUN deluser --remove-home node 

RUN apk add --no-cache git

WORKDIR /usr/src/app

RUN npm install semantic-release @semantic-release/exec

ADD semantic-release.sh ./semantic-release.sh

CMD sh semantic-release.sh
```

`semantic-release.sh`

```bash
#!/bin/sh

echo "adding user: ${USER} with UID: ${UID}"
# -D = without login password
adduser ${USER} -u ${UID} -D

echo "giving the new user ownership of application folder"
chown -R ${USER}:${USER} /usr/src/app

su ${USER} -c "npx semantic-release --no-ci --extends ./aws/semantic-release.js --repository-url ${REPO}
```

`aws/semantic-release.js`

These are my project-specific settings that I don't want to disclose here.


**Running**

```
docker build \
    -t my-semantic-release:latest .
```

```
docker run --rm -it \
    --env GH_TOKEN=$GH_TOKEN \
    --env REPO=$REPO \
    --env USER=$USER \
    --env UID=$UID \
    -v $(pwd)/aws:/usr/src/app/aws \
    -v $(pwd)/.git:/usr/src/app/.git \
    my-semantic-release:latest

```
